File: tp08/main_empty.c
#include <math.h>
#include <stdio.h>


void exercice1(void){
    printf("\n\nEXERCICE 1\n\n");
    
    char ma_chaine[] = "J'aime les pizzas"; // déclaration et affectation
    
    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/
    
    return;
}


void exercice2(void){
    printf("\n\nEXERCICE 2\n\n");
    
    char ma_chaine[] = "On est à Genève";
    
    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/
    
    return;
}


void exercice3(void) {
    printf("\n\nEXERCICE 3\n\n");

    int height = 7;
    
    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/

    return;
}

void exercice4(void) {
    printf("\n\nEXERCICE 4\n\n");

    int x = 42;
    printf("++x: %d\n", ++x);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x++: %d\n", x++);
    printf("x après %d\n\n", x);

    x = 42;
    printf("x+=1: %d\n", x += 1);
    printf("x après %d\n\n", x);

    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/
    
    return;
}

/******************** Votre fonction ci-dessous ********************/

/******************** Votre fonction ci-dessus *********************/

void exercice5(void) {
    printf("\n\nEXERCICE 5\n\n");
   
    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/

    return;
}


void exercice6(void) {
    printf("\n\nEXERCICE 6\n\n");
    
    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/

    return;
}

void exercice7(void) {
    printf("\n\nEXERCICE 7\n\n");
    
    int ouch = 3 < 4 < 2;
    printf("%d\n", ouch);

    return;
}

void exercice8(void) {
    printf("\n\nEXERCICE 8\n\n");
    
    /**************************************************************/
    /********* Exemple d'overflow du type 'unsigned char' *********/
    /**************************************************************/

    unsigned char a, b, c;

    // Vérifions qu'un 'unsigned char' occupe bien 8 bits
    int size_of_unsigned_char = sizeof(a);
    printf("Le type 'unsigned char' utilise %d octets, c'est-à-dire %d bits.\n", size_of_unsigned_char, 8 * size_of_unsigned_char);

    // Exemple d'overflow
    a = 61;
    b = 213;
    c = a + b;  // Overflow : c ne vaut pas 61 + 213 = 274 mais 274 % 256 = 18
    printf("%d + %d = %d\n", a, b, c);
    
    /*************************************/
    /********* Fin de l'exemple *********/
    /*************************************/


    /******************** Votre code ci-dessous ********************/

    /******************** Votre code ci-dessus *********************/

    return;
}


int main(void) {   

    // Astuce : commenter tous les exercices sauf celui en cours pour gagner du temps !

    exercice1();
    exercice2();
    exercice3();
    exercice4();
    exercice5();    
    exercice6();
    exercice7();
    exercice8();

    return 0;
}

File: tp03/main_empty.py
import random

from utils import exercice

#
# EXERCICE 1
#

@exercice
def exercice1():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice2():
    ma_liste = [4, 3, 7, 10, 42, 1]
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice3():
    ma_liste = [4, 3, 7, 10, 42, 1]
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice4():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

def fibo_iteratif(n):   # fonction à un seul argument
    # TODO n° 2 de l'ex 5:
    # ******************** Votre code ci-dessous ********************
    pass    # il faut des "return"
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice5():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice6():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

def tri(ma_liste):  # fonction à un seul argument
    # TODO n° 2 de l'exercice 7 (version in-place):
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice7():
    import random
    random.seed(42)
    # Liste pseudo-aléatoire avec seed fixée (la liste ne change pas si vous relancez)
    ma_liste = [random.randrange(100) for _ in range(20)]   # 20 nombres
    # Cette façon de créer des listes s'appelle "list comprehension" et s'exécute
    # généralement plus rapidement qu'une boucle for en plusieurs lignes.
    # Cependant, par soucis de compréhension, il est des fois préférable d'écrire
    # la boucle sur plusieurs lignes.
    # Le _ entre le for et le in est juste pour dire qu'on n'utilise pas cette valeur.
    print("Liste à trier:", ma_liste) 
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice8():
    import random
    random.seed(7)
    # Liste pseudo-aléatoire avec seed fixée (la liste ne change pas si vous relancez)
    ma_liste = [random.randrange(100) for _ in range(10)]   # 10 nombres
    print("Liste à séparer:", ma_liste) 
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************


@exercice
def exercice9():
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice10():
    liste_de_liste = [[93, 49, 71], [36, 83, 53], [66, 32, 51]]
    liste_applatie = []
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

# Fonction à compléter ex11:
def est_palindrome(ma_chaine):
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

@exercice
def exercice11():
    ma_chaine = input("Entrez la chaine de caractères: ")
    # ******************** Votre code ci-dessous ********************
    pass
    # ******************** Votre code ci-dessus *********************

    
if __name__ == "__main__":
    # Astuce : commenter tous les exercices sauf celui en cours pour gagner du temps !
    exercice1()
    exercice2()
    exercice3()
    exercice4()
    exercice5()
    exercice6()
    exercice7()
    exercice8()
    exercice9()
    exercice10()
    exercice11()
